ref_cache/transaction.c

```diff
@@ -0,0 +1,828 @@
+                          "%s 200 OK\r\n"
+                          "Content-Type: %s\r\n"
+                          "Content-Length: %zu\r\n"
+                          "%s\r\n",
+                          vers, content_type, len,
+                          keep_alive ? "" : "Connection: close\r\n");
+
+    if (l > text_len || text_len - l < len) { // Shouldn't happen
+        free(text);
+        set_error_response(transact, 500);
+    }
+
+    memcpy(text + l, message, len);
+
+    transact->rc = 200;
+    transact->text = text;
+    transact->sz   = l + len;
+    transact->out  = 0;
+    transact->flags &= ~TRANSACT_TEXT_CONST;
+    transact->state = TRANSACT_GOT_TEXT;
+
+    if (!keep_alive) {
+        transact->flags &= ~TRANSACT_KEEP_ALIVE;
+    }
+}
+
+static inline ssize_t send_file(Transaction *transact, int out_fd, int in_fd,
+                                off_t end) {
+#if HAVE_LINUX_SENDFILE
+    assert(end >= transact->fd_sent);
+    return sendfile(out_fd, in_fd,
+                    &transact->fd_sent, (size_t) (end - transact->fd_sent));
+
+#elif HAVE_FREEBSD_SENDFILE
+
+    off_t sbytes = 0;
+    int res = sendfile(in_fd, out_fd,
+                       transact->fd_sent, (size_t) (end - transact->fd_sent),
+                       NULL, &sbytes, 0);
+    transact->fd_sent += sbytes;
+    return res < 0 ? res : (ssize_t) sbytes;
+#elif HAVE_MACOSX_SENDFILE
+
+    off_t len = end - transact->fd_sent;
+    if (len == 0)
+      return 0; // Or the whole file gets sent
+    int res = sendfile(in_fd, out_fd, transact->fd_sent, &len, NULL, 0);
+    transact->fd_sent += len;
+    return res < 0 ? res : (ssize_t) len;
+#else
+    struct iovec iov[2];
+    int iovcnt = 1;
+    off_t count = end - transact->fd_sent;
+    unsigned int buffered;
+    ssize_t bytes = 0;
+
+    if (!transact->buffer) {
+        transact->buffer = malloc(SEND_BUFFER_SIZE);
+        if (!transact->buffer)
+            return -1;
+        transact->buffer_in = transact->buffer_out = 0;
+        transact->flags &= ~TRANSACT_SEND_BUFFER_FULL;
+    }
+
+    // Adjust count for bytes already in buffer
+    if ((transact->flags & TRANSACT_SEND_BUFFER_FULL) != 0) {
+        buffered = SEND_BUFFER_SIZE;
+    } else {
+        if (transact->buffer_out <= transact->buffer_in) {
+            buffered = transact->buffer_in - transact->buffer_out;
+        } else {
+            buffered = SEND_BUFFER_SIZE - transact->buffer_out + transact->buffer_in;
+        }
+    }
+    count -= buffered;
+    assert(count >= 0);
+
+    // Only read if there's data available and the buffer has a reasonable
+    // amount of space free (to avoid lots of short reads).
+    while (count > 0 && buffered < SEND_BUFFER_SIZE / 2) {
+        size_t len = (transact->buffer_in < transact->buffer_out
+                      ? transact->buffer_out - transact->buffer_in
+                      : SEND_BUFFER_SIZE - transact->buffer_in);
+        if ((off_t) len > count)
+            len = count;
```

Comment:
I got a warning here which this cast silences, but whether it's really necessary or the right answer… 🤷 

```suggestion
            len = (size_t) count;
```

Reply:
Yes, there's an annoying difference in the signed-ness of the parameters to `pread()`.  The cast is safe anyway, as we know in order to get here `count < len && len <= SEND_BUFFER_SIZE`.